consensus/core/src/dag_state.rs

```diff
@@ -257,6 +257,62 @@ impl DagState {
         Ok(blocks)
     }
 
+    /// Returns the last block proposed per authority with round <= `before_round`. The method will look
+    /// into cache first, otherwise will fall back in store. The method guarantees to return a result for
+    /// each authority, even if that's genesis block. In case of equivocation for an authority's last slot
+    /// only one block will be returned (the last in order).
+    pub(crate) fn get_last_block_per_authority(
+        &self,
+        before_round: Round,
+    ) -> ConsensusResult<Vec<VerifiedBlock>> {
+        let mut blocks = vec![None; self.context.committee.size()];
+
+        for (authority_index, block_refs) in self.cached_refs.iter().enumerate() {
+            if let Some(block_ref) = block_refs
+                .range((
+                    Included(BlockRef::new(0, AuthorityIndex::ZERO, BlockDigest::MIN)),
+                    Excluded(BlockRef::new(
+                        before_round + 1,
+                        AuthorityIndex::ZERO,
+                        BlockDigest::MIN,
+                    )),
+                ))
+                .next_back()
+            {
+                let block = self
+                    .recent_blocks
+                    .get(block_ref)
+                    .expect("Block should exist in recent blocks");
+
+                blocks[block_ref.author] = Some(block.clone());
+            }
+        }
+
+        // Now find out all the authority positions that are None. For those look into store
+        // If still nothing is found then return genesis. We are deliberately not filtering
+        for (authority_index, _) in self.context.committee.authorities() {
+            if blocks[authority_index].is_some() {
+                continue;
+            }
+            let result =
+                self.store
+                    .scan_last_blocks_by_author(authority_index, 1, Some(before_round))?;
+            if result.is_empty() {
+                let (_, genesis_block) = self
+                    .genesis
+                    .iter()
+                    .find(|(block_ref, _)| block_ref.author == authority_index)
+                    .expect("Genesis should be found for authority {authority_index}");
+                blocks[authority_index] = Some(genesis_block.clone());
+            } else {
+                // in case of equivocation we'll just get whatever is the last one
+                blocks[authority_index] = result.last().cloned();
```

Comment:
curious if there is any benefit of returning the block with the highest number of ancestors for potential links later